auth/impersonate/impersonate.go

```diff
@@ -0,0 +1,208 @@
+	}
+	for _, v := range opts.Delegates {
+		its.delegates = append(its.delegates, formatIAMServiceAccountName(v))
+	}
+	its.scopes = make([]string, len(opts.Scopes))
+	copy(its.scopes, opts.Scopes)
+
+	if isStaticToken {
+		return auth.NewCachedTokenProvider(its, nil), nil
+	}
+	return auth.NewCachedTokenProvider(its, nil), nil
+}
+
+// CredentialOptions for generating an impersonated credential token.
+type CredentialOptions struct {
+	// TargetPrincipal is the email address of the service account to
+	// impersonate. Required.
+	TargetPrincipal string
+	// Scopes that the impersonated credential should have. Required.
+	Scopes []string
+	// Delegates are the service account email addresses in a delegation chain.
+	// Each service account must be granted roles/iam.serviceAccountTokenCreator
+	// on the next service account in the chain. Optional.
+	Delegates []string
+	// Lifetime is the amount of time until the impersonated token expires. If
+	// unset the token's lifetime will be one hour and be automatically
+	// refreshed. If set the token may have a max lifetime of one hour and will
+	// not be refreshed. Service accounts that have been added to an org policy
+	// with constraints/iam.allowServiceAccountCredentialLifetimeExtension may
+	// request a token lifetime of up to 12 hours. Optional.
+	Lifetime time.Duration
+	// Subject is the sub field of a JWT. This field should only be set if you
+	// wish to impersonate as a user. This feature is useful when using domain
+	// wide delegation. Optional.
+	Subject string
+
+	// TokenProvider is the provider of the credentials used to fetch the ID
+	// token. If not provided, and a Client is also not provided, credentials
+	// will try to be detected from the environment. Optional.
+	TokenProvider auth.TokenProvider
+	// Client configures the underlying client used to make network requests
+	// when fetching tokens. If provided the client should provide it's own
+	// credentials at call time. Optional.
+	Client *http.Client
+}
+
+func formatIAMServiceAccountName(name string) string {
+	return fmt.Sprintf("projects/-/serviceAccounts/%s", name)
+}
+
+type generateAccessTokenReq struct {
+	Delegates []string `json:"delegates,omitempty"`
+	Lifetime  string   `json:"lifetime,omitempty"`
+	Scope     []string `json:"scope,omitempty"`
+}
+
+type generateAccessTokenResp struct {
+	AccessToken string `json:"accessToken"`
+	ExpireTime  string `json:"expireTime"`
+}
+
+type impersonatedTokenProvider struct {
+	client *http.Client
+
+	targetPrincipal string
+	lifetime        string
+	scopes          []string
+	delegates       []string
+}
+
+// Token returns an impersonated Token.
+func (i impersonatedTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+	reqBody := generateAccessTokenReq{
+		Delegates: i.delegates,
+		Lifetime:  i.lifetime,
+		Scope:     i.scopes,
+	}
+	b, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, fmt.Errorf("impersonate: unable to marshal request: %v", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("impersonate: unable to marshal request: %w", err)
```
Maybe just `s/%v", err/%w", err/` 😛 I won't bombard you with more comments.

Reply:
Thanks